File: lib/dataforseo.ts
/**
 * DataForSEO API Client
 *
 * Provides typed methods for keyword tracking, backlink analysis,
 * and on-page audits via the DataForSEO v3 REST API.
 *
 * Auth: Basic (base64-encoded login:password)
 * Docs: https://docs.dataforseo.com/v3/
 */

const BASE_URL = "https://api.dataforseo.com/v3";
const DEFAULT_DOMAIN = "theparticlepost.com";

// ---------------------------------------------------------------------------
// Response types
// ---------------------------------------------------------------------------

export interface KeywordRanking {
  keyword: string;
  position: number | null;
  url: string | null;
  searchVolume: number | null;
  cpc: number | null;
  competition: number | null;
  lastChecked: string;
}

export interface BacklinkSummary {
  totalBacklinks: number;
  referringDomains: number;
  referringIps: number;
  referringSubnets: number;
  brokenBacklinks: number;
  followLinks: number;
  nofollowLinks: number;
  domainRank: number | null;
  lastUpdated: string;
}

export interface ReferringDomain {
  domain: string;
  backlinks: number;
  firstSeen: string;
  lastSeen: string;
  rank: number | null;
  isLost: boolean;
}

export interface BrokenBacklink {
  url: string;
  anchor: string;
  referringPage: string;
  referringDomain: string;
  firstSeen: string;
  lastSeen: string;
  statusCode: number | null;
}

export interface KeywordVolume {
  keyword: string;
  searchVolume: number | null;
  cpc: number | null;
  competition: number | null;
  monthlySearches: { year: number; month: number; volume: number }[] | null;
}

export interface OnPageResult {
  taskId: string;
  status: string;
  crawledPages: number;
  totalPages: number;
  issues: OnPageIssue[];
  summary: Record<string, number>;
}

export interface OnPageIssue {
  severity: "critical" | "warning" | "info";
  type: string;
  description: string;
  pages: string[];
  count: number;
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

function getCredentials(): { login: string; password: string } {
  const login = process.env.DATAFORSEO_LOGIN;
  const password = process.env.DATAFORSEO_PASSWORD;
  if (!login || !password) {
    throw new Error(
      "DATAFORSEO_LOGIN and DATAFORSEO_PASSWORD environment variables are required"
    );
  }
  return { login, password };
}

function authHeader(): string {
  const { login, password } = getCredentials();
  const encoded = Buffer.from(`${login}:${password}`).toString("base64");
  return `Basic ${encoded}`;
}

async function apiFetch<T>(
  path: string,
  method: "GET" | "POST" = "GET",
  body?: unknown
): Promise<T> {
  const url = `${BASE_URL}${path}`;
  const res = await fetch(url, {
    method,
    headers: {
      Authorization: authHeader(),
      "Content-Type": "application/json",
    },
    ...(body ? { body: JSON.stringify(body) } : {}),
  });

  if (!res.ok) {
    const text = await res.text().catch(() => "");
    throw new Error(
      `DataForSEO API error ${res.status} ${res.statusText}: ${text}`
    );
  }

  return res.json() as Promise<T>;
}

// ---------------------------------------------------------------------------
// Generic DataForSEO envelope
// ---------------------------------------------------------------------------

interface DfsEnvelope<T> {
  version: string;
  status_code: number;
  status_message: string;
  tasks: {
    id: string;
    status_code: number;
    status_message: string;
    result: T[] | null;
  }[];
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Get SERP rankings for a list of keywords on a given domain.
 * Uses the SERP Regular POST endpoint (Google, US, English).
 */
export async function getKeywordRankings(
  keywords: string[],
  domain: string = DEFAULT_DOMAIN
): Promise<KeywordRanking[]> {
  const tasks = keywords.map((kw) => ({
    keyword: kw,
    location_code: 2840, // United States
    language_code: "en",
    device: "desktop",
    os: "windows",
    depth: 100,
  }));

  const data = await apiFetch<DfsEnvelope<SerpResult>>(
    "/serp/google/organic/live/regular",
    "POST",
    tasks
  );

  const now = new Date().toISOString();
  const results: KeywordRanking[] = [];

  for (let i = 0; i < data.tasks.length; i++) {
    const task = data.tasks[i];
    const keyword = keywords[i];
    const items = task.result?.[0]?.items ?? [];
    const match = items.find(
      (item: SerpItem) =>
        item.type === "organic" &&
        item.domain?.includes(domain)
    );

    results.push({
      keyword,
      position: match?.rank_absolute ?? null,
      url: match?.url ?? null,
      searchVolume: null,
      cpc: null,
      competition: null,
      lastChecked: now,
    });
  }

  return results;
}

/**
 * Get a backlink summary for the domain.
 */
export async function getBacklinkSummary(
  domain: string = DEFAULT_DOMAIN
): Promise<BacklinkSummary> {
  const data = await apiFetch<DfsEnvelope<BacklinkSummaryRaw>>(
    "/backlinks/summary/live",
    "POST",
    [{ target: domain, internal_list_limit: 0, include_subdomains: true }]
  );

  const r = data.tasks[0]?.result?.[0];
  if (!r) {
    throw new Error("No backlink summary data returned");
  }

  return {
    totalBacklinks: r.total_backlinks ?? 0,
    referringDomains: r.referring_domains ?? 0,
    referringIps: r.referring_ips ?? 0,
    referringSubnets: r.referring_subnets ?? 0,
    brokenBacklinks: r.broken_backlinks ?? 0,
    followLinks: r.follow ?? 0,
    nofollowLinks: r.nofollow ?? 0,
    domainRank: r.rank ?? null,
    lastUpdated: new Date().toISOString(),
  };
}

/**
 * Get referring domains for the target.
 */
export async function getReferringDomains(
  domain: string = DEFAULT_DOMAIN,
  limit: number = 100
): Promise<ReferringDomain[]> {
  const data = await apiFetch<DfsEnvelope<ReferringDomainsRaw>>(
    "/backlinks/referring_domains/live",
    "POST",
    [
      {
        target: domain,
        limit,
        include_subdomains: true,
        order_by: ["rank,desc"],
      },
    ]
  );

  const items = data.tasks[0]?.result?.[0]?.items ?? [];
  return items.map((d: ReferringDomainItem) => ({
    domain: d.domain ?? "",
    backlinks: d.backlinks ?? 0,
    firstSeen: d.first_seen ?? "",
    lastSeen: d.last_visited ?? "",
    rank: d.rank ?? null,
    isLost: d.is_lost ?? false,
  }));
}

/**
 * Start an on-page crawl/audit for the domain.
 * Returns the task ID to poll with `getOnPageResults`.
 */
export async function startOnPageAudit(
  domain: string = DEFAULT_DOMAIN
): Promise<string> {
  const data = await apiFetch<DfsEnvelope<{ id: string }>>(
    "/on_page/task_post",
    "POST",
    [
      {
        target: `https://${domain}/`,
        max_crawl_pages: 200,
        enable_javascript: true,
        load_resources: true,
        enable_browser_rendering: true,
        custom_user_agent:
          "Mozilla/5.0 (compatible; ParticlePostBot/1.0; +https://theparticlepost.com)",
      },
    ]
  );

  const taskId = data.tasks[0]?.id;
  if (!taskId) throw new Error("Failed to start on-page audit — no task ID");
  return taskId;
}

/**
 * Get the results of a previously started on-page audit.
 */
export async function getOnPageResults(
  taskId: string
): Promise<OnPageResult> {
  const data = await apiFetch<DfsEnvelope<OnPageSummaryRaw>>(
    "/on_page/summary",
    "POST",
    [{ id: taskId }]
  );

  const r = data.tasks[0]?.result?.[0];
  if (!r) {
    return {
      taskId,
      status: data.tasks[0]?.status_message ?? "unknown",
      crawledPages: 0,
      totalPages: 0,
      issues: [],
      summary: {},
    };
  }

  const issues: OnPageIssue[] = [];
  const crawlStatus = r.crawl_status;

  // Map page-level issues from the summary
  if (r.page_metrics) {
    const m = r.page_metrics;
    if (m.checks) {
      for (const [checkName, count] of Object.entries(m.checks)) {
        if (typeof count === "number" && count > 0) {
          const severity = checkName.includes("critical")
            ? "critical"
            : checkName.includes("warning")
              ? "warning"
              : "info";
          issues.push({
            severity,
            type: checkName,
            description: checkName.replace(/_/g, " "),
            pages: [],
            count,
          });
        }
      }
    }
  }

  return {
    taskId,
    status: crawlStatus.status ?? data.tasks[0]?.status_message ?? "unknown",
    crawledPages: crawlStatus.pages_crawled ?? 0,
    totalPages: crawlStatus.max_crawl_pages ?? 0,
    issues,
    summary: r.page_metrics?.checks ?? {},
  };
}

/**
 * Get broken (lost) backlinks for the domain.
 */
export async function getBrokenBacklinks(
  domain: string = DEFAULT_DOMAIN
): Promise<BrokenBacklink[]> {
  const data = await apiFetch<DfsEnvelope<BrokenBacklinksRaw>>(
    "/backlinks/backlinks/live",
    "POST",
    [
      {
        target: domain,
        mode: "as_is",
        filters: ["is_lost", "=", true],
        limit: 100,
        include_subdomains: true,
      },
    ]
  );

  const items = data.tasks[0]?.result?.[0]?.items ?? [];
  return items.map((b: BrokenBacklinkItem) => ({
    url: b.url_to ?? "",
    anchor: b.anchor ?? "",
    referringPage: b.url_from ?? "",
    referringDomain: b.domain_from ?? "",
    firstSeen: b.first_seen ?? "",
    lastSeen: b.last_seen ?? "",
    statusCode: b.page_from_status_code ?? null,
  }));
}

/**
 * Get search volume and competition data for a list of keywords.
 * Uses the Keywords Data / Google Ads endpoint.
 */
export async function getKeywordVolume(
  keywords: string[]
): Promise<KeywordVolume[]> {
  const data = await apiFetch<DfsEnvelope<KeywordVolumeRaw>>(
    "/keywords_data/google_ads/search_volume/live",
    "POST",
    [
      {
        keywords,
        location_code: 2840,
        language_code: "en",
      },
    ]
  );

  const items = data.tasks[0]?.result ?? [];
  return keywords.map((kw) => {
    const match = items.find(
      (item: KeywordVolumeItem) =>
        item.keyword?.toLowerCase() === kw.toLowerCase()
    );
    return {
      keyword: kw,
      searchVolume: match?.search_volume ?? null,
      cpc: match?.cpc ?? null,
      competition: match?.competition ?? null,
      monthlySearches:
        match?.monthly_searches?.map(
          (m: { year: number; month: number; search_volume: number }) => ({
            year: m.year,
            month: m.month,
            volume: m.search_volume,
          })
        ) ?? null,
    };
  });
}

// ---------------------------------------------------------------------------
// Raw API response shapes (internal)
// ---------------------------------------------------------------------------

interface SerpItem {
  type: string;
  rank_absolute: number;
  domain: string;
  url: string;
}

interface SerpResult {
  items: SerpItem[];
}

interface BacklinkSummaryRaw {
  total_backlinks: number;
  referring_domains: number;
  referring_ips: number;
  referring_subnets: number;
  broken_backlinks: number;
  follow: number;
  nofollow: number;
  rank: number | null;
}

interface ReferringDomainsRaw {
  items: ReferringDomainItem[];
}

interface ReferringDomainItem {
  domain: string;
  backlinks: number;
  first_seen: string;
  last_visited: string;
  rank: number | null;
  is_lost: boolean;
}

interface OnPageSummaryRaw {
  crawl_status: {
    status: string;
    pages_crawled: number;
    max_crawl_pages: number;
  };
  page_metrics: {
    checks: Record<string, number>;
  };
}

interface BrokenBacklinksRaw {
  items: BrokenBacklinkItem[];
}

interface BrokenBacklinkItem {
  url_to: string;
  anchor: string;
  url_from: string;
  domain_from: string;
  first_seen: string;
  last_seen: string;
  page_from_status_code: number | null;
}

interface KeywordVolumeRaw {
  keyword: string;
  search_volume: number | null;
  cpc: number | null;
  competition: number | null;
  monthly_searches: { year: number; month: number; search_volume: number }[];
}

type KeywordVolumeItem = KeywordVolumeRaw;

File: pipeline/tools/dataforseo.py
"""
DataForSEO Tool — CrewAI BaseTool for SEO data retrieval.

Provides keyword rankings, backlink analysis, and domain metrics
via the DataForSEO API. Uses stdlib urllib (no external deps).

Environment variables:
    DATAFORSEO_LOGIN    — DataForSEO account login (email)
    DATAFORSEO_PASSWORD — DataForSEO account password

Query format (passed as the `query` argument):
    "keywords:term1,term2"       — keyword search volume & difficulty
    "backlinks:example.com"      — backlink summary for a domain
    "audit:example.com"          — on-page SEO audit for a domain
    "ranked:example.com"         — top ranked keywords for a domain
    "competitors:example.com"    — competitor domains

Examples:
    tool._run("keywords:ai business strategy,enterprise ai")
    tool._run("backlinks:theparticlepost.com")
    tool._run("audit:theparticlepost.com")
"""

import base64
import json
import os
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen

from crewai.tools import BaseTool


_API_BASE = "https://api.dataforseo.com/v3"


class DataForSEOTool(BaseTool):
    name: str = "dataforseo_search"
    description: str = (
        "Get SEO data (keyword rankings, backlinks, domain metrics) from DataForSEO API. "
        "Input format: 'keywords:term1,term2' for search volume, "
        "'backlinks:domain' for backlink data, 'audit:domain' for on-page audit, "
        "'ranked:domain' for top keywords, 'competitors:domain' for competitors."
    )

    def _get_auth_header(self) -> str:
        """Build HTTP Basic Auth header from env vars."""
        login = os.environ.get("DATAFORSEO_LOGIN", "")
        password = os.environ.get("DATAFORSEO_PASSWORD", "")
        if not login or not password:
            raise ValueError(
                "DATAFORSEO_LOGIN and DATAFORSEO_PASSWORD environment variables required"
            )
        credentials = f"{login}:{password}"
        encoded = base64.b64encode(credentials.encode()).decode()
        return f"Basic {encoded}"

    def _api_post(self, endpoint: str, payload: list[dict]) -> dict:
        """POST to DataForSEO API and return parsed JSON response."""
        url = f"{_API_BASE}{endpoint}"
        data = json.dumps(payload).encode("utf-8")
        req = Request(url, data=data, method="POST")
        req.add_header("Authorization", self._get_auth_header())
        req.add_header("Content-Type", "application/json")

        try:
            with urlopen(req, timeout=30) as resp:
                return json.loads(resp.read().decode("utf-8"))
        except HTTPError as e:
            body = e.read().decode("utf-8", errors="replace")
            return {"error": f"HTTP {e.code}: {body[:500]}"}
        except URLError as e:
            return {"error": f"URL error: {e.reason}"}
        except Exception as e:
            return {"error": str(e)}

    def _keywords(self, terms: list[str]) -> str:
        """Get search volume and keyword difficulty."""
        payload = [{"keywords": terms, "location_code": 2840, "language_code": "en"}]
        resp = self._api_post("/keywords_data/google_ads/search_volume/live", payload)

        if "error" in resp:
            return f"DataForSEO error: {resp['error']}"

        results = []
        tasks = resp.get("tasks", [])
        for task in tasks:
            for item in task.get("result", []) or []:
                keyword = item.get("keyword", "")
                vol = item.get("search_volume", 0)
                cpc = item.get("cpc", 0)
                comp = item.get("competition", "N/A")
                comp_idx = item.get("competition_index", 0)
                results.append(
                    f"  {keyword}: vol={vol}, cpc=${cpc:.2f}, "
                    f"competition={comp} (index={comp_idx})"
                )

        if not results:
            return "No keyword data returned."
        return "KEYWORD DATA:\n" + "\n".join(results)

    def _backlinks(self, domain: str) -> str:
        """Get backlink summary for a domain."""
        payload = [{"target": domain, "limit": 10}]
        resp = self._api_post("/backlinks/summary/live", payload)

        if "error" in resp:
            return f"DataForSEO error: {resp['error']}"

        results = []
        tasks = resp.get("tasks", [])
        for task in tasks:
            for item in task.get("result", []) or []:
                results.append(
                    f"  Total backlinks: {item.get('total_backlinks', 0)}\n"
                    f"  Referring domains: {item.get('referring_domains', 0)}\n"
                    f"  Referring IPs: {item.get('referring_ips', 0)}\n"
                    f"  Domain rank: {item.get('rank', 0)}\n"
                    f"  Broken backlinks: {item.get('broken_backlinks', 0)}\n"
                    f"  Referring domains nofollow: {item.get('referring_domains_nofollow', 0)}"
                )

        if not results:
            return "No backlink data returned."
        return f"BACKLINK SUMMARY for {domain}:\n" + "\n".join(results)

    def _audit(self, domain: str) -> str:
        """Get on-page SEO audit summary."""
        payload = [{"target": domain, "max_crawl_pages": 10}]
        resp = self._api_post("/on_page/summary", payload)

        if "error" in resp:
            return f"DataForSEO error: {resp['error']}"

        results = []
        tasks = resp.get("tasks", [])
        for task in tasks:
            for item in task.get("result", []) or []:
                crawl = item.get("crawl_progress", "unknown")
                pages = item.get("pages_count", 0)
                checks = item.get("checks", {})
                results.append(
                    f"  Crawl status: {crawl}\n"
                    f"  Pages crawled: {pages}\n"
                    f"  Checks summary: {json.dumps(checks, indent=2)[:800]}"
                )

        if not results:
            return "No audit data returned."
        return f"ON-PAGE AUDIT for {domain}:\n" + "\n".join(results)

    def _ranked_keywords(self, domain: str) -> str:
        """Get top ranked keywords for a domain."""
        payload = [{"target": domain, "limit": 20, "location_code": 2840, "language_code": "en"}]
        resp = self._api_post("/dataforseo_labs/google/ranked_keywords/live", payload)

        if "error" in resp:
            return f"DataForSEO error: {resp['error']}"

        results = []
        tasks = resp.get("tasks", [])
        for task in tasks:
            for item in task.get("result", []) or []:
                total = item.get("total_count", 0)
                results.append(f"  Total ranked keywords: {total}")
                for kw_item in (item.get("items") or [])[:20]:
                    kw_data = kw_item.get("keyword_data", {})
                    keyword = kw_data.get("keyword", "")
                    ranked = kw_item.get("ranked_serp_element", {})
                    pos = ranked.get("serp_item", {}).get("rank_absolute", "?")
                    vol = kw_data.get("keyword_info", {}).get("search_volume", 0)
                    results.append(f"  #{pos} — {keyword} (vol={vol})")

        if not results:
            return "No ranked keyword data returned."
        return f"RANKED KEYWORDS for {domain}:\n" + "\n".join(results)

    def _competitors(self, domain: str) -> str:
        """Get competitor domains."""
        payload = [{"target": domain, "limit": 10, "location_code": 2840, "language_code": "en"}]
        resp = self._api_post("/dataforseo_labs/google/competitors_domain/live", payload)

        if "error" in resp:
            return f"DataForSEO error: {resp['error']}"

        results = []
        tasks = resp.get("tasks", [])
        for task in tasks:
            for item in task.get("result", []) or []:
                for comp in (item.get("items") or [])[:10]:
                    domain_name = comp.get("domain", "")
                    avg_pos = comp.get("avg_position", 0)
                    intersections = comp.get("intersections", 0)
                    visibility = comp.get("full_domain_metrics", {}).get("organic", {}).get("pos_1", 0)
                    results.append(
                        f"  {domain_name}: avg_pos={avg_pos:.1f}, "
                        f"shared_keywords={intersections}, top1_keywords={visibility}"
                    )

        if not results:
            return "No competitor data returned."
        return f"COMPETITORS for {domain}:\n" + "\n".join(results)

    def _run(self, query: str) -> str:
        """Parse query format and dispatch to the appropriate method.

        Formats:
            keywords:term1,term2
            backlinks:domain.com
            audit:domain.com
            ranked:domain.com
            competitors:domain.com
        """
        if not query or ":" not in query:
            return (
                "Invalid query format. Use one of:\n"
                "  keywords:term1,term2\n"
                "  backlinks:domain.com\n"
                "  audit:domain.com\n"
                "  ranked:domain.com\n"
                "  competitors:domain.com"
            )

        action, _, value = query.partition(":")
        action = action.strip().lower()
        value = value.strip()

        if not value:
            return f"No value provided for action '{action}'."

        try:
            if action == "keywords":
                terms = [t.strip() for t in value.split(",") if t.strip()]
                if not terms:
                    return "No keywords provided."
                return self._keywords(terms)
            elif action == "backlinks":
                return self._backlinks(value)
            elif action == "audit":
                return self._audit(value)
            elif action == "ranked":
                return self._ranked_keywords(value)
            elif action == "competitors":
                return self._competitors(value)
            else:
                return (
                    f"Unknown action '{action}'. Use: keywords, backlinks, "
                    "audit, ranked, or competitors."
                )
        except ValueError as e:
            return f"Configuration error: {e}"
        except Exception as e:
            return f"DataForSEO error: {type(e).__name__}: {e}"
